File: examples/ruby/stego_encoder_demo.rb
# frozen_string_literal: true

require_relative "../../lib/ccc/stego_encoder"

# === Claude Code 后门信息隐藏算法 — Ruby 演示 ===

encoder = StegoEncoder.new

puts "=== Claude Code 后门信息隐藏算法 - Ruby 实现 ===\n"

# 演示 1：所有 8 种状态
puts "【1. 手动编码所有 8 种状态】"
encoder.demo_all_states.each do |result|
  d = result[:decoded]
  puts "#{result[:state].ljust(16)} => #{result[:encoded]}"
  puts "  解码: 中国时区=#{d[:is_china_tz]}, 域名命中=#{d[:domain_hit]}, 关键词命中=#{d[:keyword_hit]}\n"
end

# 演示 2：自动检测
puts "【2. 自动检测模拟】"
test_urls = [
  nil,
  "https://api.openai.com/v1",
  "https://api.moonshot.ai/v1",
  "https://gateway.deepseek.com/v1",
  "https://some-proxy.cn/api",
]

test_urls.each do |url|
  result = encoder.auto_encode(proxy_url: url)
  decoded = encoder.decode(result)
  puts "URL: #{url || '(无代理)'}"
  puts "  生成: #{result}"
  puts "  解码: 中国时区=#{decoded[:is_china_tz]}, 域名命中=#{decoded[:domain_hit]}, 关键词命中=#{decoded[:keyword_hit]}\n"
end

# 演示 3：XOR 混淆
puts "【3. XOR 混淆演示 (key=91)】"
original = "moonshot.ai|deepseek.com|zhipu.ai"
obfuscated = encoder.xor_obfuscate(original)
recovered = encoder.xor_deobfuscate(obfuscated)
puts "原始:    #{original}"
puts "混淆后:  #{obfuscated}"
puts "恢复:    #{recovered}"
puts "匹配验证: #{original == recovered ? 'PASS' : 'FAIL'}"

# 演示 4：Unicode 同形字符
puts "\n【4. Unicode 同形字符对比】"
chars = [
  { name: "标准 ASCII 撇号", code: 0x0027 },
  { name: "右单引号", code: 0x2019 },
  { name: "修饰字母撇号", code: 0x02BC },
  { name: "修饰字母上标号", code: 0x02B9 },
]

chars.each do |c|
  ch = c[:code].chr(Encoding::UTF_8)
  puts "U+#{c[:code].to_s(16).upcase.rjust(4, '0')}  #{c[:name].ljust(16)}  =>  \"#{ch}\"  (ord: #{ch.ord})"
end

# 演示 5：防御检测
puts "\n【5. 防御检测 - 扫描异常 Unicode】"
suspicious_prompt = "Today's date is 2026/06/30."
suspicious_prompt["'"] = "\u2019"  # 替换为右单引号
puts "测试提示词: #{suspicious_prompt}"
puts "扫描结果:"
suspicious_prompt.each_char.with_index do |ch, i|
  cp = ch.ord
  if [0x2019, 0x02BC, 0x02B9].include?(cp)
    puts "  发现可疑字符 U+#{cp.to_s(16).upcase.rjust(4, '0')} 在位置 #{i}: '#{ch}'"
  end
end

File: examples/js/stego_encoder_demo.js
const { StegoEncoder } = require("../../src/stego_encoder");

// === Claude Code 后门信息隐藏算法 — JS 演示 ===

const encoder = new StegoEncoder();

console.log("=== Claude Code 后门信息隐藏算法 - JS 实现 ===\n");

// 演示 1：手动编码各种状态
console.log("【1. 手动编码所有 8 种状态】");
const states = [
  { name: "非中国/未命中", tz: false, dom: false, kw: false },
  { name: "非中国/命中域名", tz: false, dom: true, kw: false },
  { name: "非中国/命中关键词", tz: false, dom: false, kw: true },
  { name: "非中国/两者命中", tz: false, dom: true, kw: true },
  { name: "中国/未命中", tz: true, dom: false, kw: false },
  { name: "中国/命中域名", tz: true, dom: true, kw: false },
  { name: "中国/命中关键词", tz: true, dom: false, kw: true },
  { name: "中国/两者命中", tz: true, dom: true, kw: true },
];

states.forEach(s => {
  const result = encoder.encode(s.tz, s.dom, s.kw);
  const decoded = encoder.decode(result);
  console.log(`${s.name.padEnd(16)} => ${result}`);
  console.log(`  解码: 中国时区=${decoded.isChinaTZ}, 域名命中=${decoded.domainHit}, 关键词命中=${decoded.keywordHit}\n`);
});

// 演示 2：自动检测（模拟）
console.log("【2. 自动检测模拟】");
const testUrls = [
  null,
  "https://api.openai.com/v1",
  "https://api.moonshot.ai/v1",
  "https://gateway.deepseek.com/v1",
  "https://some-proxy.cn/api",
];

testUrls.forEach(url => {
  const result = encoder.autoEncode(url);
  const decoded = encoder.decode(result);
  console.log(`URL: ${url || "(无代理)"}`);
  console.log(`  生成: ${result}`);
  console.log(`  解码: 中国时区=${decoded.isChinaTZ}, 域名命中=${decoded.domainHit}, 关键词命中=${decoded.keywordHit}\n`);
});

// 演示 3：XOR 混淆
console.log("【3. XOR 混淆演示 (key=91)】");
const original = "moonshot.ai|deepseek.com|zhipu.ai";
const obfuscated = encoder.xorObfuscate(original);
const recovered = encoder.xorDeobfuscate(obfuscated);
console.log(`原始:    ${original}`);
console.log(`混淆后:  ${obfuscated}`);
console.log(`恢复:    ${recovered}`);
console.log(`匹配验证: ${original === recovered ? "PASS" : "FAIL"}`);

// 演示 4：Unicode 同形字符对比
console.log("\n【4. Unicode 同形字符对比】");
const chars = [
  { name: "标准 ASCII 撇号", code: 0x0027 },
  { name: "右单引号", code: 0x2019 },
  { name: "修饰字母撇号", code: 0x02BC },
  { name: "修饰字母上标号", code: 0x02B9 },
];

chars.forEach(c => {
  const ch = String.fromCodePoint(c.code);
  console.log(`U+${c.code.toString(16).toUpperCase().padStart(4, "0")}  ${c.name.padEnd(16)}  =>  "${ch}"  (codePoint: ${ch.codePointAt(0)})`);
});
